src/components/tl/mlx5/mcast/tl_mlx5_mcast_coll.c

```diff
@@ -6,17 +6,249 @@
 
 #include "tl_mlx5_coll.h"
 #include "tl_mlx5_mcast_helper.h"
+#include "tl_mlx5_mcast_rcache.h"
 
-ucc_status_t ucc_tl_mlx5_mcast_test(ucc_tl_mlx5_mcast_coll_req_t* req /* NOLINT */)
+static inline ucc_status_t ucc_tl_mlx5_mcast_r_window_recycle(ucc_tl_mlx5_mcast_coll_comm_t *comm,
+                                                              ucc_tl_mlx5_mcast_coll_req_t  *req)
 {
-    return UCC_ERR_NOT_SUPPORTED;
+    ucc_status_t     status        = UCC_OK;
+    int              wsize         = comm->wsize;
+    int              num_free_win  = wsize - (comm->psn - comm->last_acked);
+    int              req_completed = (req->to_send == 0 && req->to_recv == 0);
+    struct pp_packet *pp           = NULL;
+
+    assert(comm->recv_drop_packet_in_progress == false);
+    assert(req->to_send >= 0);
+
+    /* When do we need to perform reliability protocol:
+       1. Always in the end of the window
+       2. For the zcopy case: in the end of collective, because we can't signal completion
+       before made sure that children received the data - user can modify buffer */
+
+    assert(num_free_win >= 0);
+
+    if (!num_free_win || (req->proto == MCAST_PROTO_ZCOPY && req_completed)) {
+
+        status = ucc_tl_mlx5_mcast_reliable(comm);
+        if (UCC_OK != status) {
+            return status;
+        }
+
+        comm->n_mcast_reliable++;
+
+        for (;comm->last_acked < comm->psn; comm->last_acked++) {
+            pp = comm->r_window[comm->last_acked & (wsize-1)];
+            assert(pp != &comm->dummy_packet);
+            comm->r_window[comm->last_acked & (wsize-1)] = &comm->dummy_packet;
+
+            pp->context = 0;
+            ucc_list_add_tail(&comm->bpool, &pp->super);
+        }
+
+        if (!req_completed) {
+			status = ucc_tl_mlx5_mcast_prepare_reliable(comm, req, req->root);
+            if (ucc_unlikely(UCC_OK != status)) {
+                return status;
+            }
+        }
+    }
+
+    return UCC_OK;
+}
+
+static inline ucc_status_t ucc_tl_mlx5_mcast_do_bcast(ucc_tl_mlx5_mcast_coll_req_t *req)
```

Comment:
not sure if making this function inline helps

Reply:
can you clarify further? This func is our progress func and i wanted to avoid extra calls